lib/pbench/server/cache_manager.py

```diff
@@ -1507,6 +1521,58 @@ def _discover_controllers(self):
             if file.is_dir() and file.name != CacheManager.TEMPORARY:
                 self._add_controller(file)
 
+    def _discover_datasets(self):
+        """Discover the ARCHIVE tree from the SQL representation of datasets
+
+        Discover all controllers and tarballs with server.tarball-path
+        metadata. This will find only correct and "live" tarballs, and may
+        miss incorrectly labeled tarballs. (Unless we're trying to diagnose
+        ARCHIVE tree errors, this is probably what we want.)
+        """
+        rows = (
+            Database.db_session.query(
+                Dataset.name,
+                Dataset.resource_id,
+                Metadata.value["tarball-path"].as_string(),
+            )
+            .execution_options(stream_results=True)
+            .outerjoin(
+                Metadata,
+                and_(Dataset.id == Metadata.dataset_ref, Metadata.key == "server"),
+            )
+            .yield_per(500)
+            .all()
+        )
+
+        for name, resource_id, path in rows:
+            if not all((resource_id, path)):
```

Comment:
Why check for the `resource_id` but not the `name`?  (It seems like the only item at question is the `path`, so why not check just that?)

Reply:
The name is only used for a log message, but both the `resource_id` and `path` are critical. Yeah, technically we can't have a `Dataset` without a `resource_id` (the column isn't nullable), but as long as I was checking I figured I'd check both just because the code looked odd checking just one.

Reply:
Unfortunately, the code also looks odd checking two (I had to think carefully about how the `not` applied to `any()` and why there was an "extra" set of parens in the argument list...all of which go away if you only check `path`).